client/core/core.go

```diff
@@ -2762,22 +2835,43 @@ func handleMatchRoute(c *Core, dc *dexConnection, msg *msgjson.Message) error {
 	if err != nil {
 		return fmt.Errorf("match request parsing error: %v", err)
 	}
+
+	// Acknowledgements MUST be in the same orders as the msgjson.Matches.
 	matches, acks, err := dc.parseMatches(msgMatches, true)
 	if err != nil {
+		// Even one failed match fails them all since the server requires acks
+		// for them all, and in the same order. TODO: consider lifting this
+		// requirement, which requires changes to the server's handling.
 		return err
 	}
-
 	resp, err := msgjson.NewResponse(msg.ID, acks, nil)
 	if err != nil {
 		return err
 	}
+
+	// Send the match acknowledgments.
+	// TODO: Consider a "QueueSend" or similar, but do not bail on the matches.
 	err = dc.Send(resp)
 	if err != nil {
-		return err
+		log.Errorf("Send match response: %v", err)
+		// dc.addPendingSend(resp) // e.g.
 	}
 
-	defer c.refreshUser()
+	defer c.refreshUser() // update WalletState
 
+	// Begin match negotiation.
+	//
+	// TODO: We need to know that the server has first recorded the ack
+	// signature, otherwise 'init' will fail. Until this is solved with a
+	// response to the ack, sleep for a moment if we are the maker on any of the
+	// orders. However, resendPendingRequests will retry the request when the
+	// trackedTrack ticks again when the timer fires in listen.
+	for _, m := range msgMatches {
+		if m.Side == uint8(order.Maker) {
+			time.Sleep(250 * time.Millisecond)
```

Comment:
Another use case for a response to a response, or a more general "dialog" or sequence of responses.  Even if the match ack was successfully sent, `runMatches` and init broadcast/request can happen immediately for a maker's match.  This is a race with the server's handling of the match ack because until it stores the ack signature it won't be ready for the init!  It's not a terrible problem though because `resendPendingRequests` will keep trying if it does fail because of this.

I'm now trying to mitigate this race by locking the matches mutex early in `(*Swapper).processMatchAcks` (https://github.com/decred/dcrdex/pull/565/files#diff-5f1d20749025f318cd1a8ba0915ebd37R2110-R2126), but it's still a race AFAICT.  The best solution would be for the client to get confirmation that the match ack was received and recorded before attempting the init.